File: adapter/harness/world_model.py
"""
World model data structures — P0.2.

Enforces the core invariant: observations and beliefs are separate typed
structures. A belief cannot be created without an explicit derived_from[]
chain — auto-promotion from observation is a structural error.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from typing import Any, Literal


@dataclass
class Observation:
    id: str
    content: str
    source: str
    recorded_at: datetime = field(default_factory=lambda: datetime.now(UTC))

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "content": self.content,
            "source": self.source,
            "recorded_at": self.recorded_at.isoformat(),
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Observation:
        return cls(
            id=d["id"],
            content=d["content"],
            source=d["source"],
            recorded_at=datetime.fromisoformat(d["recorded_at"]) if "recorded_at" in d else datetime.now(UTC),
        )


@dataclass
class Belief:
    id: str
    statement: str
    confidence: float
    derived_from: list[str]
    supporting_evidence: list[str] = field(default_factory=list)
    reliability: str = ""
    recorded_at: datetime = field(default_factory=lambda: datetime.now(UTC))
    # Belief ids this belief has been found to contradict — stamped by
    # WorldModel.add_contradiction() below, not set directly. Alongside derived_from[]
    # this makes the provenance graph the critique asked for ("Belief B17, derived from
    # E42/E51, contradicts B13") queryable from the belief itself, not only recoverable
    # by scanning world_model.contradictions[].involved_belief_ids for this id.
    contradicts: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "statement": self.statement,
            "confidence": self.confidence,
            "derived_from": self.derived_from,
            "supporting_evidence": self.supporting_evidence,
            "recorded_at": self.recorded_at.isoformat(),
            "contradicts": self.contradicts,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Belief:
        return cls(
            id=d["id"],
            statement=d["statement"],
            confidence=d["confidence"],
            derived_from=d["derived_from"],
            supporting_evidence=d.get("supporting_evidence", []),
            recorded_at=datetime.fromisoformat(d["recorded_at"]) if "recorded_at" in d else datetime.now(UTC),
            contradicts=d.get("contradicts", []),
        )


ContradictionType = Literal["pairwise", "set-level", "temporal", "abstraction"]
ContradictionSeverity = Literal["LOW", "MEDIUM", "HIGH", "SYSTEM_BREAKING"]
ContradictionScope = Literal["local", "task", "global"]


@dataclass
class Contradiction:
    id: str
    type: ContradictionType
    severity: ContradictionSeverity
    scope: ContradictionScope
    involved_belief_ids: list[str] = field(default_factory=list)
    # Matches TS's Contradiction.description (packages/harness/src/state/world-model.ts) — was
    # missing here entirely; defaulted to "" (not None) so every existing to_dict()/from_dict()
    # round-trip and internal detector constructor call stays byte-compatible without needing to
    # touch each one. Populated for record_external_contradiction() (Phase 2 of
    # plans/lexical_functions_hardening_plan.html), which needs somewhere to put the semantic
    # check's own finding text — the four internal lexical detectors don't set this yet, a known,
    # separate gap from that same file's TS/Python description-text drift, not fixed here.
    description: str = ""

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "type": self.type,
            "severity": self.severity,
            "scope": self.scope,
            "involved_belief_ids": self.involved_belief_ids,
            "description": self.description,
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> Contradiction:
        return cls(
            id=d["id"],
            type=d["type"],
            severity=d["severity"],
            scope=d["scope"],
            involved_belief_ids=d.get("involved_belief_ids", []),
            description=d.get("description", ""),
        )


@dataclass
class WorldModel:
    generation_id: int = 0
    observations: list[Observation] = field(default_factory=list)
    beliefs: list[Belief] = field(default_factory=list)
    assumptions: list[str] = field(default_factory=list)
    contradictions: list[Contradiction] = field(default_factory=list)
    environment_change_log: list[dict[str, Any]] = field(default_factory=list)
    completeness_flags: dict[str, bool] = field(default_factory=dict)
    # Belief IDs flagged as stale after a constraint/criteria change (P7)
    stale_flags: dict[str, bool] = field(default_factory=dict)

    def add_observation(self, obs: Observation) -> None:
        self.observations.append(obs)

    def add_belief(self, belief: Belief) -> None:
        if not belief.derived_from:
            raise ValueError(
                f"derived_from must be non-empty (belief id='{belief.id}'). "
                "A belief requires at least one source in derived_from[] — "
                "auto-promotion from an observation without an explicit derivation chain is disallowed."
            )
        self.beliefs.append(belief)

    def add_contradiction(self, contradiction: Contradiction) -> None:
        # SYSTEM_BREAKING contradictions are stored here and picked up by the
        # next resolve_control_state() Tier 1 pass — they never cause an inline
        # halt or raise an exception at this layer.
        self.contradictions.append(contradiction)
        # Stamp contradicts[] onto every involved belief so the relationship is
        # queryable from the belief itself (see Belief.contradicts' own docstring).
        # Pairwise between all involved beliefs, not just "each points at the
        # contradiction" — a 3+-way set-level contradiction means every belief in it
        # contradicts every other one, not just the contradiction record.
        beliefs_by_id = {b.id: b for b in self.beliefs}
        for belief_id in contradiction.involved_belief_ids:
            belief = beliefs_by_id.get(belief_id)
            if belief is None:
                continue
            for other_id in contradiction.involved_belief_ids:
                if other_id != belief_id and other_id not in belief.contradicts:
                    belief.contradicts.append(other_id)

    def to_dict(self) -> dict[str, Any]:
        return {
            "generation_id": self.generation_id,
            "observations": [o.to_dict() for o in self.observations],
            "beliefs": [b.to_dict() for b in self.beliefs],
            "assumptions": list(self.assumptions),
            "contradictions": [c.to_dict() for c in self.contradictions],
            "environment_change_log": list(self.environment_change_log),
            "completeness_flags": dict(self.completeness_flags),
            "stale_flags": dict(self.stale_flags),
        }

    @classmethod
    def from_dict(cls, d: dict[str, Any]) -> WorldModel:
        wm = cls(
            generation_id=d.get("generation_id", 0),
            assumptions=d.get("assumptions", []),
            environment_change_log=d.get("environment_change_log", []),
            completeness_flags=d.get("completeness_flags", {}),
            stale_flags=d.get("stale_flags", {}),
        )
        for o in d.get("observations", []):
            wm.observations.append(Observation.from_dict(o))
        for b in d.get("beliefs", []):
            # Bypass the validation guard when deserialising — the data was
            # already validated when it was first persisted.
            wm.beliefs.append(Belief.from_dict(b))
        for c in d.get("contradictions", []):
            wm.contradictions.append(Contradiction.from_dict(c))
        return wm

File: packages/harness/src/state/world-model.ts
import { z } from 'zod'

export const ObservationSchema = z.object({
  id: z.string(),
  content: z.string(),
  source: z.string(),
  timestamp: z.string(),
})
export type Observation = z.infer<typeof ObservationSchema>

export const BeliefSchema = z.object({
  id: z.string(),
  content: z.string(),
  reliability: z.enum(['HIGH', 'MEDIUM', 'LOW']),
  derived_from: z.array(z.string()),
  timestamp: z.string(),
})
export type Belief = z.infer<typeof BeliefSchema>

export const ContradictionSchema = z.object({
  id: z.string(),
  type: z.enum(['pairwise', 'set-level', 'temporal', 'abstraction']),
  severity: z.enum(['LOW', 'MEDIUM', 'HIGH', 'SYSTEM_BREAKING']),
  scope: z.enum(['local', 'task', 'global']),
  description: z.string(),
  belief_ids: z.array(z.string()),
})
export type Contradiction = z.infer<typeof ContradictionSchema>

export const EnvironmentChangeSchema = z.object({
  id: z.string(),
  description: z.string(),
  affected_paths: z.array(z.string()),
  timestamp: z.string(),
})
export type EnvironmentChange = z.infer<typeof EnvironmentChangeSchema>

export const BeliefNodeSchema = z.object({
  belief_id: z.string(),
  confidence: z.number(),
})
export type BeliefNode = z.infer<typeof BeliefNodeSchema>

export const DepEdgeSchema = z.object({
  from: z.string(),
  to: z.string(),
  confidence: z.number(),
  verified: z.boolean(),
})
export type DepEdge = z.infer<typeof DepEdgeSchema>

export const PropagationTaskSchema = z.object({
  source_belief_id: z.string(),
  target_belief_id: z.string(),
})
export type PropagationTask = z.infer<typeof PropagationTaskSchema>

export const BeliefDepGraphSchema = z.object({
  belief_nodes: z.array(BeliefNodeSchema),
  derived_from_edges: z.array(DepEdgeSchema),
  invalidation_frontier: z.array(z.string()),
  propagation_queue: z.array(PropagationTaskSchema),
  unverified_edge_ratio: z.number(),
  confidence_decay_rate: z.number(),
})
export type BeliefDepGraphData = z.infer<typeof BeliefDepGraphSchema>

export class BeliefDepGraph {
  belief_nodes: BeliefNode[]
  derived_from_edges: DepEdge[]
  invalidation_frontier: string[]
  propagation_queue: PropagationTask[]
  unverified_edge_ratio: number
  confidence_decay_rate: number

  constructor(data?: Partial<BeliefDepGraphData>) {
    this.belief_nodes = data?.belief_nodes ?? []
    this.derived_from_edges = data?.derived_from_edges ?? []
    this.invalidation_frontier = data?.invalidation_frontier ?? []
    this.propagation_queue = data?.propagation_queue ?? []
    this.unverified_edge_ratio = data?.unverified_edge_ratio ?? 0
    this.confidence_decay_rate = data?.confidence_decay_rate ?? 0.05
  }

  recomputeUnverifiedEdgeRatio(): void {
    const total = this.derived_from_edges.length
    if (total === 0) { this.unverified_edge_ratio = 0; return }
    const unverified = this.derived_from_edges.filter(e => !e.verified).length
    this.unverified_edge_ratio = unverified / total
  }

  toJSON(): BeliefDepGraphData {
    return {
      belief_nodes: this.belief_nodes,
      derived_from_edges: this.derived_from_edges,
      invalidation_frontier: this.invalidation_frontier,
      propagation_queue: this.propagation_queue,
      unverified_edge_ratio: this.unverified_edge_ratio,
      confidence_decay_rate: this.confidence_decay_rate,
    }
  }

  static fromJSON(json: BeliefDepGraphData): BeliefDepGraph {
    const parsed = BeliefDepGraphSchema.parse(json)
    return new BeliefDepGraph(parsed)
  }
}

export const DepGraphBudgetSchema = z.object({
  max_unverified_edge_ratio: z.number(),
  refresh_policy: z.string(),
  confidence_decay_rate: z.number(),
})
export type DepGraphBudgetData = z.infer<typeof DepGraphBudgetSchema>

export class DepGraphBudget {
  max_unverified_edge_ratio: number
  refresh_policy: string
  confidence_decay_rate: number

  constructor(data?: Partial<DepGraphBudgetData>) {
    this.max_unverified_edge_ratio = data?.max_unverified_edge_ratio ?? 0.3
    this.refresh_policy = data?.refresh_policy ?? 'per_iteration'
    this.confidence_decay_rate = data?.confidence_decay_rate ?? 0.05
  }

  applyDecay(graph: BeliefDepGraph): void {
    for (const edge of graph.derived_from_edges) {
      edge.confidence = Math.max(0, edge.confidence - this.confidence_decay_rate)
    }
    graph.recomputeUnverifiedEdgeRatio()
  }

  toJSON(): DepGraphBudgetData {
    return {
      max_unverified_edge_ratio: this.max_unverified_edge_ratio,
      refresh_policy: this.refresh_policy,
      confidence_decay_rate: this.confidence_decay_rate,
    }
  }
}

export const WorldModelSchema = z.object({
  generation_id: z.number().int().nonnegative(),
  observations: z.array(ObservationSchema),
  beliefs: z.array(BeliefSchema),
  assumptions: z.array(z.string()),
  contradictions: z.array(ContradictionSchema),
  environment_change_log: z.array(EnvironmentChangeSchema),
  completeness_flags: z.record(z.boolean()),
})
export type WorldModelData = z.infer<typeof WorldModelSchema>

export class WorldModel {
  generation_id: number
  observations: Observation[]
  beliefs: Belief[]
  assumptions: string[]
  contradictions: Contradiction[]
  environment_change_log: EnvironmentChange[]
  completeness_flags: Record<string, boolean>

  constructor(data?: Partial<WorldModelData>) {
    this.generation_id = data?.generation_id ?? 0
    this.observations = data?.observations ?? []
    this.beliefs = data?.beliefs ?? []
    this.assumptions = data?.assumptions ?? []
    this.contradictions = data?.contradictions ?? []
    this.environment_change_log = data?.environment_change_log ?? []
    this.completeness_flags = data?.completeness_flags ?? {}
  }

  addBelief(belief: Belief): void {
    if (!belief.derived_from || belief.derived_from.length === 0) {
      throw new Error(
        `WorldModel invariant: belief "${belief.id}" must have a non-empty derived_from[] chain`,
      )
    }
    this.beliefs.push(belief)
  }

  incrementGenerationId(): void {
    this.generation_id++
  }

  toJSON(): WorldModelData {
    return {
      generation_id: this.generation_id,
      observations: this.observations,
      beliefs: this.beliefs,
      assumptions: this.assumptions,
      contradictions: this.contradictions,
      environment_change_log: this.environment_change_log,
      completeness_flags: this.completeness_flags,
    }
  }

  static fromJSON(json: WorldModelData): WorldModel {
    const parsed = WorldModelSchema.parse(json)
    return new WorldModel(parsed)
  }
}
